tsl/src/compression/compress_utils.c

```diff
@@ -564,3 +510,120 @@ tsl_decompress_chunk(PG_FUNCTION_ARGS)
+	InitFunctionCallInfoData(**decompfn_fcinfo,
+							 decompfn,
+							 2,
+							 InvalidOid, /* collation */
+							 cxt,
+							 NULL);
+	FC_ARG(*decompfn_fcinfo, 0) = FC_ARG(orig_fcinfo, 0);
+	FC_NULL(*decompfn_fcinfo, 0) = FC_NULL(orig_fcinfo, 0);
+	FC_ARG(*decompfn_fcinfo, 1) = FC_ARG(orig_fcinfo, 1);
+	FC_NULL(*decompfn_fcinfo, 1) = FC_NULL(orig_fcinfo, 1);
+}
+
+static Datum
+tsl_recompress_remote_chunk(Chunk *uncompressed_chunk, FunctionCallInfo fcinfo, bool if_compressed)
+{
+	FmgrInfo decompfn;
+	FmgrInfo compfn;
+	FunctionCallInfo decompfn_fcinfo;
+	FunctionCallInfo compfn_fcinfo;
+	get_compression_fcinfo(DECOMPRESS_CHUNK_FUNCNAME, &decompfn, &decompfn_fcinfo, fcinfo);
+
+	FunctionCallInvoke(decompfn_fcinfo);
+	if (decompfn_fcinfo->isnull)
+	{
+		ereport((if_compressed ? NOTICE : ERROR),
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("decompression failed for chunk \"%s\"",
+						get_rel_name(uncompressed_chunk->table_id)),
+				 errdetail("The compression status for the chunk is %d",
+						   uncompressed_chunk->fd.status)));
+
+		PG_RETURN_NULL();
+	}
+	get_compression_fcinfo(COMPRESS_CHUNK_FUNCNAME, &compfn, &compfn_fcinfo, fcinfo);
+	Datum compoid = FunctionCallInvoke(compfn_fcinfo);
+	if (compfn_fcinfo->isnull)
+	{
+		ereport((if_compressed ? NOTICE : ERROR),
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("compression failed for chunk \"%s\"",
+						get_rel_name(uncompressed_chunk->table_id)),
+				 errdetail("The compression status for the chunk is %d",
+						   uncompressed_chunk->fd.status)));
+		PG_RETURN_NULL();
+	}
+	return compoid;
+}
+
+bool
+tsl_recompress_chunk_wrapper(Chunk *uncompressed_chunk)
+{
+	Oid uncompressed_chunk_relid = uncompressed_chunk->table_id;
+	if (ts_chunk_is_unordered(uncompressed_chunk))
+	{
+		if (!decompress_chunk_impl(uncompressed_chunk->hypertable_relid,
+								   uncompressed_chunk_relid,
+								   false))
+			return false;
+	}
+	Chunk *chunk = ts_chunk_get_by_relid(uncompressed_chunk_relid, true);
+	Assert(!ts_chunk_is_compressed(chunk));
+	tsl_compress_chunk_wrapper(chunk, false);
+	return true;
+}
+
+Datum
+tsl_recompress_chunk(PG_FUNCTION_ARGS)
+{
+	Oid uncompressed_chunk_id = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	bool if_compressed = PG_ARGISNULL(1) ? false : PG_GETARG_BOOL(1);
+	Chunk *uncompressed_chunk = ts_chunk_get_by_relid(uncompressed_chunk_id, true);
+	if (uncompressed_chunk == NULL)
+		elog(ERROR, "unknown chunk id %d", uncompressed_chunk_id);
+	if (!ts_chunk_is_unordered(uncompressed_chunk))
+	{
+		if (!ts_chunk_is_compressed(uncompressed_chunk))
+		{
+			ereport((if_compressed ? NOTICE : ERROR),
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+					 errmsg("call compress_chunk instead of recompress_chunk")));
+			PG_RETURN_NULL();
+		}
+		else
+		{
+			ereport((if_compressed ? NOTICE : ERROR),
+					(errcode(ERRCODE_DUPLICATE_OBJECT),
+					 errmsg("chunk \"%s\" is already compressed",
+							get_rel_name(uncompressed_chunk->table_id))));
```

Comment:
I think this error can be confusing. When recompressing, you'd expect the chunk to be compressed already, so why is this an error?  Also, the error code doesn't seem appropriate given that you'd expect this object to exist when recompressing, so there is no duplicate object. Maybe you mean to say something like:

```suggestion
					(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
					 errmsg("nothing to recompress in chunk \"%s\"",
							get_rel_name(uncompressed_chunk->table_id))));
```

Reply:
fixed.